x/twap/logic_test.go

```diff
@@ -1596,3 +1596,129 @@ func geometricTestCaseFromDeltas0(startAccum, accumDiff sdk.Dec, timeDelta time.
 	return geometricTestCaseFromDeltas0(startAccum, accumDiff, timeDelta, sdk.OneDec().Quo(expectedTwap))
 }
+
+func (s *TestSuite) TestTwapLog() {
+	var expectedErrTolerance = osmomath.MustNewDecFromStr("0.000000000000000100")
+
+	tests := map[string]struct {
+		priceValue    osmomath.BigDec
+		expected      osmomath.BigDec
+		expectedPanic bool
+	}{
+		"TwapLog(-1); invalid; panic": {
+			priceValue:    osmomath.OneDec().Neg(),
+			expectedPanic: true,
+		},
+		"TwapLog(0); invalid; panic": {
+			priceValue:    osmomath.ZeroDec(),
+			expectedPanic: true,
+		},
+		"TwapLog(0.001) = -9.965784284662087044": {
+			priceValue: osmomath.MustNewDecFromStr("0.001"),
+			// From:  https://www.wolframalpha.com/input?i2d=true&i=log+base+2+of0.001with+19+digits
+			expected: osmomath.MustNewDecFromStr("-9.965784284662087044"),
+		},
+		"TwapLog(0.05) = -4.3219280948873623479": {
+			priceValue: osmomath.MustNewDecFromStr("0.05"),
+			// From:  https://www.wolframalpha.com/input?i2d=true&i=log+base+2+of+0.05+with+20+digits
+			expected: osmomath.MustNewDecFromStr("-4.3219280948873623479"),
+		},
+		"TwapLog(1) = 0": {
+			priceValue: osmomath.OneDec(),
+			expected:   osmomath.ZeroDec(),
+		},
+		"TwapLog(2) = 1": {
+			priceValue: osmomath.NewBigDec(2),
+			expected:   osmomath.OneDec(),
+		},
+		"TwapLog(13) = 3.700439718141092160": {
+			priceValue: osmomath.NewBigDec(13),
+			// From: https://www.wolframalpha.com/input?i2d=true&i=log+base+2+of+13+with+19+digits
+			expected: osmomath.MustNewDecFromStr("3.700439718141092160"),
+		},
+		"TwapLog(128) = 7": {
+			priceValue: osmomath.NewBigDec(128),
+			expected:   osmomath.NewBigDec(7),
+		},
+		"TwapLog(1024) = -4.3219280948873623479": {
+			priceValue: osmomath.MustNewDecFromStr("0.05"),
+			// From:  https://www.wolframalpha.com/input?i2d=true&i=log+base+2+of+0.05+with+20+digits
+			expected: osmomath.MustNewDecFromStr("-4.3219280948873623479"),
+		},
+		"Twaplog{1024.987654321} = 10.00139081765414132": {
+			priceValue: osmomath.NewDecWithPrec(1024987654321, 9),
+			// From: https://www.wolframalpha.com/input?i2d=true&i=log+base+2+of+1024.987654321+with+19+digits
+			expected: osmomath.MustNewDecFromStr("10.00139081765414132"),
+		},
+		"Twaplog{912648174127941279170121098210.928219201902041311} = 99.525973560175362367": {
+			priceValue: osmomath.MustNewDecFromStr("912648174127941279170121098210.928219201902041311"),
+			// From: https://www.wolframalpha.com/input?i2d=true&i=log+base+2+of+912648174127941279170121098210.928219201902041311+with+20+digits
+			expected: osmomath.MustNewDecFromStr("99.525973560175362367"),
+		},
+	}
+	for name, test := range tests {
+		s.Run(name, func() {
+			osmoassert.ConditionalPanic(s.T(), test.expectedPanic, func() {
+				result := twap.TwapLog(test.priceValue.SDKDec())
+				require.True(osmomath.DecApproxEq(s.T(), test.expected, osmomath.BigDecFromSDKDec(result), expectedErrTolerance))
```

Comment:
We need to make sure that we are testing that the following value is used for the base: `geometricTwapMathBase`. This would also require two simple tests case as opposed to many table driven.

That would imply exporting it for tests via `export_test.go`.

Test case 1:
- Performing `twap.Log` and `osmomath.CustomLog` with `geometricTwapMathBase` and comparing the values.

Test case 2: using one of the hardcoded values above. For example, `Twaplog{912648174127941279170121098210.928219201902041311} = 99.525973560175362367`

No need to keep other test cases to avoid increasing LOC. The underlying log implementation should be testing this thoroughly

Reply:
I have updated the code with your suggestions about comparing values, can you plz check again?

Reply:
Nice work!